atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java

```diff
@@ -463,6 +474,189 @@ public String toString() {
+        Set<Entry<InetSocketAddress, List<byte[]>>> rowsByHost =
+                partitionByHost(rows, Functions.<byte[]>identity()).entrySet();
+        List<Callable<Map<byte[], RowColumnRangeIterator>>> tasks = Lists.newArrayListWithCapacity(rowsByHost.size());
+        for (final Map.Entry<InetSocketAddress, List<byte[]>> hostAndRows : rowsByHost) {
+            tasks.add(() ->
+                    getRowsColumnRangeIteratorForSingleHost(hostAndRows.getKey(), tableRef,
+                            hostAndRows.getValue(), columnRangeSelection, timestamp));
+        }
+        List<Map<byte[], RowColumnRangeIterator>> perHostResults = runAllTasksCancelOnFailure(tasks);
+        Map<byte[], RowColumnRangeIterator> result = Maps.newHashMapWithExpectedSize(Iterables.size(rows));
+        for (Map<byte[], RowColumnRangeIterator> perHostResult : perHostResults) {
+            result.putAll(perHostResult);
+        }
+        return result;
+    }
+
+    private Map<byte[], RowColumnRangeIterator> getRowsColumnRangeIteratorForSingleHost(InetSocketAddress host,
+                                                                                        TableReference tableRef,
+                                                                                        List<byte[]> rows,
+                                                                                        ColumnRangeSelection columnRangeSelection,
+                                                                                        long startTs) {
+        try {
+            RowColumnRangeExtractor.RowColumnRangeResult firstPage =
+                    getRowsColumnRangeForSingleHost(host, tableRef, rows, columnRangeSelection, startTs);
+
+            Map<byte[], LinkedHashMap<Cell, Value>> results = firstPage.results;
+            Map<byte[], Column> rowsToLastCompositeColumns = firstPage.rowsToLastCompositeColumns;
+            Map<byte[], byte[]> incompleteRowsToNextColumns = Maps.newHashMap();
+            for (byte[] row : rowsToLastCompositeColumns.keySet()) {
+                byte[] col = CassandraKeyValueServices.decomposeName(rowsToLastCompositeColumns.get(row)).getLhSide();
+                // If we read a version of the cell before our start timestamp, it will be the most recent version readable to
+                // us and we can continue to the next column. Otherwise we have to continue reading this column.
+                boolean completedCell = results.containsKey(row) && results.get(row).containsKey(Cell.create(row, col));
+                boolean endOfRange = isEndOfColumnRange(completedCell, col, columnRangeSelection.getEndCol());
+                if (!endOfRange) {
+                    byte[] nextCol = getNextColumnRangeColumn(completedCell, col);
+                    incompleteRowsToNextColumns.put(row, nextCol);
+                }
+            }
+
+            Map<byte[], RowColumnRangeIterator> ret = Maps.newHashMapWithExpectedSize(rows.size());
```

Comment:
wasn't planning on doing that since we don't in getRows, but we could if it's worth the extra work to keep them ordered after merging all our cassandra results.